app/controllers/api/messages_controller.rb

```diff
@@ -0,0 +1,152 @@
+        format.xml
+        format.json
+      end
+    end
+
+    # Create a new message from current user
+    def create
+      # Check the arguments are sane
+      raise OSM::APIBadUserInput, "No title was given" if params[:title].blank?
+      raise OSM::APIBadUserInput, "No body was given" if params[:body].blank?
+      raise OSM::APIBadUserInput, "No body format was given" if params[:body_format].blank?
+
+      # Extract the arguments
+      if params[:recipient_id]
+        recipient_id = params[:recipient_id].to_i
+        recipient = User.find(recipient_id)
+      elsif params[:recipient]
+        recipient_display_name = params[:recipient]
+        recipient = User.find_by(:display_name => recipient_display_name)
+      else
+        raise OSM::APIBadUserInput, "No recipient was given"
+      end
+      title = params[:title]
+      body = params[:body]
+      body_format = params[:body_format]
+
+      raise OSM::APIRateLimitExceeded if current_user.sent_messages.where("sent_on >= ?", Time.now.utc - 1.hour).count >= current_user.max_messages_per_hour
+
+      @message = Message.new(:sender => current_user,
+                             :recipient => recipient,
+                             :sent_on => Time.now.utc,
+                             :title => title,
+                             :body => body,
+                             :body_format => body_format)
+      @message.save!
+
+      UserMailer.message_notification(@message).deliver_later if @message.notify_recipient?
+
+      # Return a copy of the new message
+      respond_to do |format|
+        format.xml { render :action => :show }
+        format.json { render :action => :show }
+      end
+    end
+
+    # Update read status of a message
+    def update
+      @message = Message.find(params[:id])
+      read_status_idx = %w[true false].index params[:read_status]
+
+      raise OSM::APIBadUserInput, "Invalid value of `read_status` was given" if read_status_idx.nil?
+      raise OSM::APIAccessDenied unless current_user.id == @message.to_user_id
+
+      @message.message_read = read_status_idx.zero?
+      @message.save!
+
+      # Return a copy of the message
+      respond_to do |format|
+        format.xml { render :action => :show }
+        format.json { render :action => :show }
+      end
+    end
+
+    # Delete message by marking it as not visible for the current user
+    def destroy
+      @message = Message.find(params[:id])
+      if current_user.id == @message.from_user_id
+        @message.from_user_visible = false
+      elsif current_user.id == @message.to_user_id
+        @message.to_user_visible = false
+      else
+        raise OSM::APIAccessDenied
+      end
+
+      @message.save!
+
+      # Return a copy of the message
+      respond_to do |format|
+        format.xml { render :action => :show }
+        format.json { render :action => :show }
+      end
+    end
+
+    private
+
+    def show_messages
+      if params[:min_id]
```

Comment:
Elsewhere (#4532) I wanted to standardize on `before`/`after` for exclusive bounds (already in use in pagination) and `from`/`to` fro inclusive bounds (already in use in the api, although just for dates).

Reply:
Using timestamp in this case would require us to add an additional index. There is also an edge case of multiple messages with same timestamp, potentially resulting in result set longer than "limit".

I wonder if it is possible for the list to be mis-ordered by timestamp, when it is ordered by `id`? I guess it depends on how autoincrement is implemented.

I am not sure if adding an index is worth it...

Reply:
I'm not saying you should sort by timestamps. I'm saying that instead of `min` I'd call this parameter `from`, because that's how it's already called in other places. But since in other places it's for dates and here it's for ids, I guess it should be `from_id` or maybe `from[id]`.

Reply:
I am not sure about this - how would the request for latest messages look like with `to` or `to_id`?

get 10 messages in descending order, then get the following 10 messages in desc order (assuming the smallest id from 1st request was 1000):
```
GET /api/0.6/user/messages/inbox?to=0&limit=10
GET /api/0.6/user/messages/inbox?to=1000&limit=10
```

instead of:
```
GET /api/0.6/user/messages/inbox?max_id=0&limit=10
GET /api/0.6/user/messages/inbox?max_id=1000&limit=10
```

Here `max_id` would imply descending, it wouldn't be so obvious if the name of parameter is named `to` or `to_id`.

Reply:
That's because you change the order based on these parameters. Elsewhere in the api, the order is either always descending or there's a parameter to choose between descending/ascending.

Reply:
Maybe you can argue that `min` and `max` look better than `from` and `to`. In this case we should add them as aliases for `from`/`to` for other endpoints.

Reply:
`from_id` and `order=newest` or `order=oldest` would be more in line with existing OSM api?